app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt

```diff
@@ -1,165 +1,292 @@
+        ) : ParserError() {
+            override val errorMessage =
+                if (symbol != null) "The symbol <<$symbol>> is unexpected at location $location."
+                else "The end of file is unexpected."
         }
 
-        data class PositionalArgumentAfterNamed(val functionCall: Expression.FunctionCall) : ArgumentResolutionError() {
-            override fun toString(): String = "Positional argument after a named one in function call $functionCall"
-        }
-
-        data class MissingArgument(val functionCall: Expression.FunctionCall, val argumentName: String) : ArgumentResolutionError() {
-            override fun toString(): String = "Argument $argumentName in function call $functionCall"
-        }
-
-        data class RepeatedArgument(val functionCall: Expression.FunctionCall, val argumentName: String) : ArgumentResolutionError() {
-            override fun toString(): String = "Argument $argumentName is repeated in function call $functionCall"
-        }
-
-        data class UnknownArgument(val functionCall: Expression.FunctionCall, val argumentName: String) : ArgumentResolutionError() {
-            override fun toString(): String = "Unknown named argument $argumentName in function call $functionCall"
-        }
-
-        data class TooManyArguments(val functionCall: Expression.FunctionCall) : ArgumentResolutionError() {
-            override fun toString(): String = "Too many arguments in function call $functionCall"
+        class InvalidNumberLiteral(
+            number: String,
+            location: LocationRange,
+        ) : ParserError() {
+            override val errorMessage = "The literal <<$number>> at location $location is an invalid number literal."
         }
     }
 
-    sealed class TypeCheckingError : Diagnostic() {
-        override fun isError() = true
-
-        data class ConstantWithoutValue(val variable: Variable) : TypeCheckingError() {
-            override fun toString() = "A constant must have a value"
-        }
-
-        data class UninitializedGlobalVariable(val variable: Variable) : TypeCheckingError() {
-            override fun toString() = "A global variable must be initialized"
-        }
-
-        data class ImmutableAssignment(val assignment: Statement.Assignment, val variable: Variable) : TypeCheckingError() {
-            override fun toString() = "Cannot assign to a " + if (variable.kind == Variable.Kind.CONSTANT) "constant" else "value"
-        }
-
-        data class ParameterAssignment(val assignment: Statement.Assignment, val parameter: Function.Parameter) : TypeCheckingError() {
-            override fun toString() = "Cannot assign to a parameter"
-        }
-
-        data class FunctionAssignment(val assignment: Statement.Assignment, val function: Function) : TypeCheckingError() {
-            override fun toString() = "Cannot assign to a function"
-        }
+    sealed class ResolutionError(astNodes: List<AstNode>) : Diagnostic {
```

Comment:
The argument probably would be more convenient as `vararg AstNode`.

Reply:
I tried and I had a problem with passing it to the child class (some typing error).